bpf: implement unconnected udp based host lb

borkmann opened the pull request:
This gets nslookup/dig working for unconnected UDP over IPv4/IPv6.

For details, see commit message.

(Requires kernel fix and iproute2/bpftool modification before we can push to Cilium repo.)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8100)
<!-- Reviewable:end -->

borkmann committed:
bpf: implement unconnected udp based host lb

This gets nslookup, dig and friends working for UDP over IPv4/IPv6.

Idea is to have a LRU map as a cache only in kernel space where
requests are tracked on a per socket basis, aka socket cookie as
part of the lookup key. It's large enough that this should be
sufficient for current connections, old entries are purged as
per LRU. Stale service entries are detected and automatically
removed in the LRU if there is a revnat mismatch or reverse lookup
fails.

One acceptable limitation is that if a socket does a direct request
to a backend instead of using the service IP (and the service IP
still exists in the service lookup table), then it may get reverse
translated wrongly due to the cached entry.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

vadorovsky commented on bpf/bpf_sock.c:
```diff
@@ -95,6 +106,105 @@ int sock4_xlate(struct bpf_sock_addr *ctx)
+	};
+
+	rval = map_lookup_elem(&cilium_cgroups_revnat4, &rkey);
+	if (!rval) {
```
Maybe

```
if (rval) {
    struct lb4_service_v2 *svc;
    ...
}

return RECVMSG_PROCEED;
```
?

vadorovsky commented on bpf/bpf_sock.c:
```diff
@@ -150,6 +259,106 @@ int sock6_xlate(struct bpf_sock_addr *ctx)
+	ctx_get_v6_address(ctx, &rkey.address);
+
+	rval = map_lookup_elem(&cilium_cgroups_revnat6, &rkey);
+	if (!rval) {
```
Maybe

```
if (rval) {
    struct lb6_service_v2 *svc;
    ...
}

return RECVMSG_PROCEED;
```
?

vadorovsky reviewed (commented):
One minor nit.

brb committed:
WIP: test host-lb

- Use cilium/ubuntu-next:v57 which runs the latest bpf tree with the recv
  hook.
- Enable host-lb by default.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
